pkg/kubelet/volume/cache/actual_state_of_world.go

```diff
@@ -0,0 +1,622 @@
+}
+
+func (asw *actualStateOfWorld) AddVolume(
+	volumeSpec *volume.Spec) (api.UniqueVolumeName, error) {
+	asw.Lock()
+	defer asw.Unlock()
+
+	volumePlugin, err := asw.volumePluginMgr.FindPluginBySpec(volumeSpec)
+	if err != nil || volumePlugin == nil {
+		return "", fmt.Errorf(
+			"failed to get Plugin from volumeSpec for volume %q err=%v",
+			volumeSpec.Name(),
+			err)
+	}
+
+	volumeName, err :=
+		volumehelper.GetUniqueVolumeNameFromSpec(volumePlugin, volumeSpec)
+	if err != nil {
+		return "", fmt.Errorf(
+			"failed to GetUniqueVolumeNameFromSpec for volumeSpec %q using volume plugin %q err=%v",
+			volumeSpec.Name(),
+			volumePlugin.GetPluginName(),
+			err)
+	}
+
+	pluginIsAttachable := false
+	if _, ok := volumePlugin.(volume.AttachableVolumePlugin); ok {
+		pluginIsAttachable = true
+	}
+
+	volumeObj, volumeExists := asw.attachedVolumes[volumeName]
+	if !volumeExists {
+		volumeObj = attachedVolume{
+			volumeName:         volumeName,
+			spec:               volumeSpec,
+			mountedPods:        make(map[volumetypes.UniquePodName]mountedPod),
+			pluginName:         volumePlugin.GetPluginName(),
+			pluginIsAttachable: pluginIsAttachable,
+			globallyMounted:    false,
+		}
+		asw.attachedVolumes[volumeName] = volumeObj
+	}
+
+	return volumeObj.volumeName, nil
+}
+
+func (asw *actualStateOfWorld) AddPodToVolume(
+	podName volumetypes.UniquePodName,
+	podUID types.UID,
+	volumeName api.UniqueVolumeName,
+	mounter volume.Mounter,
+	outerVolumeSpecName string,
+	volumeGidValue string) error {
+	asw.Lock()
+	defer asw.Unlock()
+
+	volumeObj, volumeExists := asw.attachedVolumes[volumeName]
+	if !volumeExists {
+		return fmt.Errorf(
+			"no volume with the name %q exists in the list of attached volumes",
+			volumeName)
+	}
+
+	podObj, podExists := volumeObj.mountedPods[podName]
+	if !podExists {
+		podObj = mountedPod{
+			podName:             podName,
+			podUID:              podUID,
+			mounter:             mounter,
+			outerVolumeSpecName: outerVolumeSpecName,
+			volumeGidValue:      volumeGidValue,
+		}
+	}
+
+	// If pod exists, reset remountRequired value
+	podObj.remountRequired = false
+	asw.attachedVolumes[volumeName].mountedPods[podName] = podObj
+
+	return nil
+}
+
+func (asw *actualStateOfWorld) MarkRemountRequired(
+	podName volumetypes.UniquePodName) {
+	asw.Lock()
+	defer asw.Unlock()
+	for volumeName, volumeObj := range asw.attachedVolumes {
+		for mountedPodName, podObj := range volumeObj.mountedPods {
+			if mountedPodName != podName {
+				continue
+			}
+
+			volumePlugin, err :=
+				asw.volumePluginMgr.FindPluginBySpec(volumeObj.spec)
+			if err != nil || volumePlugin == nil {
+				// Log and continue processing
+				glog.Errorf(
+					"MarkRemountRequired failed to FindPluginBySpec for pod %q (podUid %q) volume: %q (volSpecName: %q)",
+					podObj.podName,
+					podObj.podUID,
+					volumeObj.volumeName,
+					volumeObj.spec.Name())
+				continue
+			}
+
+			if volumePlugin.RequiresRemountOnPodUpdate() {
```

Comment:
I'm a little confused about this -- it makes it look like we will only ever remount volumes that return true for `RequiresRemountOnPodUpdate`